# DATA BIAS PROJECT

### Exploring the concept of bias through querying the Perspective API released by Google Jigsaw. In this project I specifically analyze the hidden biases that affect the scoring of religious terms. 

In [27]:
import pandas as pd
import numpy as np
import time

df = pd.read_csv('/Users/audreydahlkemper/Downloads/labeled_and_scored_comments.csv')


In [2]:
conda install -c conda-forge google-api-python-client

Solving environment: done

## Package Plan ##

  environment location: /Users/audreydahlkemper/opt/anaconda3

  added / updated specs:
    - google-api-python-client


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-22.9.0               |   py39h6e9494a_1         969 KB  conda-forge
    google-api-python-client-1.12.8|     pyhd3deb0d_0          48 KB  conda-forge
    google-auth-httplib2-0.1.0 |     pyhd8ed1ab_1          13 KB  conda-forge
    httplib2-0.20.4            |     pyhd8ed1ab_0          96 KB  conda-forge
    python_abi-3.9             |           2_cp39           4 KB  conda-forge
    simplejson-3.17.6          |   py39h63b48b0_1         102 KB  conda-forge
    uritemplate-3.0.1          |             py_0          16 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.2 MB

The fol

In [10]:
df.sort_values(['score'])

id                                       comment_text  \
1158   61a85266de70cca0                         Можешь говорить по русски.   
5246   030a9ba0b32b4456  Rodzina Krusenstern Potomkowie Admirała Adama ...   
11040  a6814193ac6b4705  The opening paragraph states that the game is ...   
24192  f0d6dee96e057a5b  The book is used as a source in a number of ar...   
2332   cbb9268d3c7429c1     Vals Vienés \n |Rumba \n |Cha Cha Chá(ciclo 7)   
...                 ...                                                ...   
19994  1429a2c930688317                          you fucking piece of shit   
28728  d82e45eb103ad07b                   fuck off you stupid aspy asshole   
25535  e9b92c6c6b01aeef  FUCK YOU YOU PIECE OF SHIT \n FUCK YOU YOU PIE...   
25945  e783fd267f3a9d3b  FUCK WIKIPEDIA ON WHEELS! \n\nFuck off wikiped...   
17833  364b6c24210e2345  Fuck you you fucking pig!\n\nYou motherfucking...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
1158   0.000004      0             0        0       0       0              0  
5246   0.000009      0             0        0       0       0              0  
11040  0.000091      0             0        0       0       0              0  
24192  0.000116      0             0        0       0       0              0  
2332   0.000131      0             0        0       0       0              0  
...         ...    ...           ...      ...     ...     ...            ...  
19994  0.997278      1             1        1       0       1              0  
28728  0.997736      1             1        1       0       1              0  
25535  0.997982      1             1        1       0       1              0  
25945  0.998136      1             1        1       0       1              0  
17833  0.998329      1             0        1       0       1              0  

[41338 rows x 9 columns]

In [20]:
from googleapiclient.discovery import build
import json

def get_toxicity_score(comment):
    
    API_KEY = 'AIzaSyClhOaJW7IZtxkzsAVEUtXMaqR0zlD749U'
    
    client = build(
    "commentanalyzer",
    "vlalphal",
    developerKey = API_KEY,
    discoveryServiceUrl = "https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1" 
    )
    
    analyze_request = {
    'comment': {'text': comment}, 
    'requestedAttributes': {'TOXICITY': {}}
    }
    
    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]
    
    return toxicity_score

In [33]:
threshold = 0.5

df['predicition'] = (df['score']> threshold).astype(int)
df['predicition'].value_counts()

0    34925
1     6413
Name: predicition, dtype: int64

In [34]:
from sklearn.metrics import confusion_matrix

In [35]:
confusion_matrix(df['toxic'], df['predicition'])

array([[34677,  2718],
       [  248,  3695]])

** Testing Different Toxicity Scores

In [36]:
get_toxicity_score('kill')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.47762036, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.47762036, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.47762036

In [37]:
get_toxicity_score('banned')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 6, 'score': {'value': 0.0789077, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.0789077, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.0789077

In [38]:
get_toxicity_score('religion')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 8, 'score': {'value': 0.026722329, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.026722329, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.026722329

In [40]:
get_toxicity_score('mangina')

HttpError: <HttpError 400 when requesting https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=AIzaSyClhOaJW7IZtxkzsAVEUtXMaqR0zlD749U&alt=json returned "Attribute TOXICITY does not support request languages: mg". Details: "[{'@type': 'type.googleapis.com/google.commentanalyzer.v1alpha1.Error', 'errorType': 'LANGUAGE_NOT_SUPPORTED_BY_ATTRIBUTE', 'languageNotSupportedByAttributeError': {'detectedLanguages': ['mg'], 'attribute': 'TOXICITY'}}]">

I found the above toxicity score search intersting. After skimming the labeled and scored comments csv I found a section where a user had repeatidly used this word in reference to somebody. This is not a word that is found in the common venacular but it a hybrid of two words (man and vagina). It is frequently used to emmasculate men and is drawn from the broad cultural belief systems that label females as weak and men as strong. The programs inability to label 'mangina' with a toxicity score' represents a possible flaw within the system as it may be unable to keep up with changing language patterns. 

In [52]:
get_toxicity_score('I LOVE DICK MORE THAN ANYTHING')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 30, 'score': {'value': 0.8346757, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.8346757, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.8346757

In [53]:
get_toxicity_score('I')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 1, 'score': {'value': 0.018048171, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.018048171, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.018048171

In [54]:
get_toxicity_score('love')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.024147147, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.024147147, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.024147147

In [55]:
get_toxicity_score('dick')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.59863794, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.59863794, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.59863794

In [56]:
get_toxicity_score('more')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.044748597, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.044748597, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.044748597

In [57]:
get_toxicity_score('than')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.019019997, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.019019997, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.019019997

In [59]:
get_toxicity_score('anything')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 8, 'score': {'value': 0.022945397, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.022945397, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.022945397

'I love dick more than anything' also proved to be interesting. The phrase generated a score of 0.83. When each individual word was parsed the only variable that returned a high score was 'dick' with 0.59. Dick is interesting because it referes to both an insult and to a part of bodily anatomy. This could be a statment of great love, an expression of sexuality. It seems that the program is translating this statment as homophobic which may be influenced by the demographic that frequents chat pages/online groups - primarily straight, white, men who use these forums to reassert their masculinity which in western society is an ideology that opposes homosexuality. 

In [63]:
get_toxicity_score('I love vagina more than anything')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 32, 'score': {'value': 0.8269591, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.8269591, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.8269591

In [64]:
get_toxicity_score('I love pussy more than anything')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 31, 'score': {'value': 0.83020335, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.83020335, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.83020335

In [60]:
get_toxicity_score('vagina')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 6, 'score': {'value': 0.5042868, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.5042868, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.5042868

In [61]:
get_toxicity_score('pussy')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.6955873, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.6955873, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.6955873

In [62]:
get_toxicity_score('cunt')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.5026305, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.5026305, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.5026305

Did a comparison of female anatomy. The sentances 'I love pussy more than anything' and 'I love vagina more than anything' generated a similar toxicity score to 'I love dick more than anything'. But the words 'vagina' and 'cunt' were lower in toxicity while 'pussy' ranked higher. 

In [49]:
get_toxicity_score("Info-Nazi's")

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 11, 'score': {'value': 0.34388447, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.34388447, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en', 'hr']}


0.34388447

In [50]:
get_toxicity_score('loser')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.67902285, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.67902285, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.67902285

'loser' generated a high toxicity score - 0.67. Something to think about. Perhaps it scored so high because it is generaly applicable to everybody and commonly used. 

In [65]:
get_toxicity_score('bible')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.022773718, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.022773718, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.022773718

In [66]:
get_toxicity_score('quran')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.028954152, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.028954152, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en', 'ms']}


0.028954152

In [67]:
get_toxicity_score('torah')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.029984225, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.029984225, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.029984225

In [68]:
get_toxicity_score('christian')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 9, 'score': {'value': 0.027580723, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.027580723, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.027580723

In [69]:
get_toxicity_score('muslim')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 6, 'score': {'value': 0.16275369, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.16275369, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.16275369

In [70]:
get_toxicity_score('jewish')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 6, 'score': {'value': 0.2712817, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.2712817, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.2712817

In [173]:
get_toxicity_score('jews')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 4, 'score': {'value': 0.24418297, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.24418297, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.24418297

Another potential bias within the system. References to christianity ranked lower on the toxicity score. Bible was set at 0.022, Quran 0.028, and Torah 0.029. The differences between religious texts references were less but when the actual religion was referenced the results were more varied. Christian was 0.027, muslim was 0.162, and jewish was 0.271. Why is one relgion considered more toxic then others? 

In [42]:
get_toxicity_score('bible thumping')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 14, 'score': {'value': 0.29855597, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.29855597, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.29855597

In [51]:
get_toxicity_score('beat the bible')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 14, 'score': {'value': 0.42390585, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.42390585, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.42390585

Another interesting change in toxicity score was noted for the two variables above - 'bible thumping' vs 'beat the bible'. Both are used in reference to a certain demographic with the insinuation of overreliance on a specific text for general knowledge. However despite both colloqualisms referring to the same group, when the phrase was reworded it generated a higher toxicity score. 'Bible thumping' was 0.29 while 'beat the bible' was 0.42. 

In [44]:
get_toxicity_score('stereotype')

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 10, 'score': {'value': 0.055333257, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.055333257, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}


0.055333257

## Multi-Variable Comparison and Hypothesis Description

# # # Hypothesis: The Perspective API scores the religious terms related to Christianity, Islam, and Judiaism differently because of the unconcious socio-cultural biases of the programmers or the code that the descriptors were sourced from. Western society holds a pro-white, Christian perspective which has caused Christianity related terms to score lower on the Toxicity scale and Jewish and Islamic related terms to score higher on the Toxicity scale. 

The first table compares the toxicity scores generated by the names of the three largest religions as well as several variations of each sect: Christian/Christianity/Christians, Islam/Islamic/Muslim/Muslims, Israeli/Judaism/Jewish/Jews.

In [169]:
comment_list = ['christian', 'christianity', 'christians', 'Islam', 'Islamic','muslim', 'muslims', 'israeli', 'judaism','jewish', 'jews']
score_list = []

for comment in comment_list:
    score = get_toxicity_score(comment)
    score_list.append(score)
    print(comment, score)
    time.sleep(1)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 9, 'score': {'value': 0.027580723, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.027580723, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
christian 0.027580723
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 12, 'score': {'value': 0.03753809, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.03753809, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
christianity 0.03753809
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 10, 'score': {'value': 0.059728492, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.059728492, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
christians 0.059728492
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.055333257, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.055333257, 'type': 'PROBABILITY'}}}, '

In [170]:
sum(score_list)

1.324451062

In [171]:
religions= pd.DataFrame({'comment': comment_list, 'score': score_list})

In [140]:
religions.describe()

score
count  11.000000
mean    0.120405
std     0.088902
min     0.027581
25%     0.049698
50%     0.069718
75%     0.181424
max     0.271282

In [141]:
religions.head(11)

comment     score
0      christian  0.027581
1   christianity  0.037538
2     christians  0.059728
3          Islam  0.055333
4        Islamic  0.152178
5         muslim  0.162754
6        muslims  0.200095
7        israeli  0.044062
8        judaism  0.069718
9         jewish  0.271282
10          jews  0.244183

The highest toxicity score was 'jewish' with 0.271. The lowest score related to Judaism was 'israeli' with 0.044. That score was slightly lower than the highest toxicity score for words related to Christianity. The highest toxicity score for Christianity was 'christians' with 0.059. None of the words associated with Christianity returned a score higher than 0.059. The highest score for Islam related words was 'muslims' with 0.200 and the lowest was 'Islam' with 0.055. On average, the word that directly ties to the religious sect (Christian 0.027, Islam 0.055, Judaism 0.069) scored lower then the variation of the word used to describe the people praciticing the related relgion - especially when the people were referred to as a group rather then an individual - (Christians 0.059, Muslims 0.200, Jewish/Jews 0.271/0.244). 

The second table analyzes the toxicity scores for each religions primary religious book: Bible, Qur'an, and Torah. This also includes two related words - scripture and scroll - which could reference multiple religous sects. 

In [149]:
comment_list = ['bible', 'quran', 'torah', 'scripture', 'scroll']
score_list = []

for comment in comment_list:
    score = get_toxicity_score(comment)
    score_list.append(score)
    print(comment, score)
    time.sleep(1)

{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.022773718, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.022773718, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
bible 0.022773718
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.028954152, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.028954152, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en', 'ms']}
quran 0.028954152
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 5, 'score': {'value': 0.029984225, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.029984225, 'type': 'PROBABILITY'}}}, 'languages': ['en'], 'detectedLanguages': ['en']}
torah 0.029984225
{'attributeScores': {'TOXICITY': {'spanScores': [{'begin': 0, 'end': 9, 'score': {'value': 0.035477944, 'type': 'PROBABILITY'}}], 'summaryScore': {'value': 0.035477944, 'type': 'PROBABILITY'}}}, 'languages

In [150]:
sum(score_list)

0.166402282

In [151]:
book = pd.DataFrame({'comment': comment_list, 'score': score_list})

In [152]:
book.describe()

score
count  5.000000
mean   0.033280
std    0.009983
min    0.022774
25%    0.028954
50%    0.029984
75%    0.035478
max    0.049212

In [153]:
book.head()

comment     score
0      bible  0.022774
1      quran  0.028954
2      torah  0.029984
3  scripture  0.035478
4     scroll  0.049212

The words related to religious books were ranked lower on the toxicity scale then the words describing the relgion or the people in the religion. However, despite the lower toxicity score overall the Torah and Qur'an were still ranked as more toxic than the bible. It is important to note that the score differentiating the Bible, Qur'an, and Torah were seperated by 0.001 of a point. Interestingly, 'scroll' was ranked the highest of all at 0.049. 

## Analyzing the Comments Related to the Religious Terms

1. Calling forward comments with the term bible

In [131]:
df['has Bible'] = df['comment_text'].apply(lambda x: ('bible' in x.lower()))
bible_comments = df[df['has Bible']== True]
bible_comments

id                                       comment_text  \
119    07dd9cd6c4166da7  == New American Bible == \n\n Great job on the...   
547    2be5edd340b92044  2009 (UTC) \n\n That is incorrect about the Bi...   
597    2fc65f2713b28cec  == St James' Presbyterian Church of England, B...   
673    378101a3f4f6d557  " \n ::::I answered your question on his talk ...   
1781   995bf984f347e221  " \n\n ===God Stuff=== \n It looks like we're ...   
...                 ...                                                ...   
38284  a4bf4d7f4af2d321  "\n\n Creation and Destruction \n\nI wanted to...   
38520  a373db3fda857516  "\n\n Oregon Vortex \n\n Hello Queen Cake!  Yo...   
38696  a27965af095fc72d  "\n\nRequest for Help, please\nGreetings Carlo...   
40699  97cf5c4bd1fa3f02  "\n\n Biography assessment rating comment \nWi...   
40976  966929908249fc81  Maryester what long angry salad of letters tha...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
119    0.658780      0             0        0       0       0              0   
547    0.092102      0             0        0       0       0              0   
597    0.076315      0             0        0       0       0              0   
673    0.742851      0             0        0       0       0              0   
1781   0.114054      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
38284  0.455820      0             0        0       0       0              0   
38520  0.411799      1             0        0       0       0              0   
38696  0.115180      0             0        0       0       0              0   
40699  0.288563      0             0        0       0       0              0   
40976  0.136296      0             0        0       0       0              0   

       predicition  has Bible  
119              1       True  
547              0       True  
597              0       True  
673              1       True  
1781             0       True  
...            ...        ...  
38284            0       True  
38520            0       True  
38696            0       True  
40699            0       True  
40976            0       True  

[86 rows x 11 columns]

I noticed a correlation between America and the bible in a few of the text descriptions. The comments commonly cover Bible passages, the validity of the Bible, and its history. Interstingly, the bible is referrred to as its own entity in some of the lines. I also noticed that when the Bible or Christianity is brought up in an argument it's system of belief is heavily defended and attacks on the other persons character are brought up. See comment 8c86d7f9799ba2ee. 

2. Calling forward comments with the word Qur'an.

In [133]:
df['has Quran'] = df['comment_text'].apply(lambda x: ('quran' in x.lower()))
quran_comments = df[df['has Quran']== True]
quran_comments

id                                       comment_text  \
1096   5a213ffb16971b6b  == Confusion == \n I am a bit confused about t...   
1903   a18fe3306cd15eb2  " \n\n == Needs sourcing == \n The article say...   
2068   b127e10b96aab262  " \n\n == Wikipedia's Neutrality in Religion? ...   
2372   cf0224c5f58cee01  ED PLEASE.  We have had this argument over and...   
4525   8d5e84944314f6eb  The Digital Quran is a consumer product genre,...   
5045   a7080acd0610fcbb  Those are interested can check these sites: \n...   
7644   dd6e9f4dd38685dd  "Forgetting about the in-depth meaning of the ...   
9620   bd751f5e85917b55  Ah Yahel very honest of you to remove my comme...   
16282  510a0209577244de  Harun Yahya/Adnan Oktar \n\nThis is quoted fro...   
16518  4d60b95dcc7d72d2  "\n\nHi Padma and Nat!\nPadma, I can understan...   
17518  3b7913c3b973fb1a  "\n\n POV tag unwarranted \n\nCiting Quran pas...   
19183  20f92ea8e1690ba8  "\n\n Urgent sabotage of page! \n\nSignificant...   
24766  edc540ced55a5a66  "length of Quran==\ncan somebody tell me, how ...   
25636  e932d1bbeb4c3369  "\n\nReply\n Sunnah.com  is cited to cite Tirm...   
26011  e715d021c42e6d61  It is true that I can not read urdu but this i...   
26573  e3f91a794d353039  "\n\n User: Vaishnava on the Article: Vishnu \...   
28335  da67f2e4e75b5ae8  "\n\nRe the following passage: ""Christians ha...   
28898  d7436faad4e2f4a5  "\n\n confused \n\n""The Quran, however, provi...   
38725  a25b0db361a8a1d4  ok, i am sorry, but I do not agree with what i...   
40384  99a9ef39896a4724  this Baboon43 is from Scarborough, Toronto, Ca...   
40386  99a757a90e3edcba  "\n QURAN \nthis page is supposed to be the ""...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1096   0.534903      0             0        0       0       0              0   
1903   0.350313      0             0        0       0       0              0   
2068   0.315989      0             0        0       0       0              0   
2372   0.671703      0             0        0       0       0              0   
4525   0.149319      0             0        0       0       0              0   
5045   0.149300      0             0        0       0       0              0   
7644   0.282473      0             0        0       0       0              0   
9620   0.534903      0             0        0       0       0              0   
16282  0.695427      0             0        1       0       0              0   
16518  0.183096      0             0        0       0       0              0   
17518  0.209429      0             0        0       0       0              0   
19183  0.556825      0             0        0       0       0              0   
24766  0.150651      0             0        0       0       0              0   
25636  0.294464      0             0        0       0       0              0   
26011  0.182593      0             0        0       0       0              0   
26573  0.160907      0             0        0       0       0              0   
28335  0.240593      0             0        0       0       0              0   
28898  0.115797      0             0        0       0       0              0   
38725  0.278894      0             0        0       0       0              0   
40384  0.457961      0             0        0       0       0              0   
40386  0.446716      0             0        0       0       0              0   

       predicition  has Bible  has Quran  
1096             1      False       True  
1903             0      False       True  
2068             0       True       True  
2372             1      False       True  
4525             0      False       True  
5045             0      False       True  
7644             0      False       True  
9620             1      False       True  
16282            1      False       True  
16518            0      False       True  
17518            0      False       True  
19183            1      Fal

Not that many mentions of the Qua'ran in this text. Most of the comments referred to the Qur'an for reasons related to historical analysis or religious comparisons. 

3. Calling forward comments containing the word 'Torah'.

In [135]:
df['has Torah'] = df['comment_text'].apply(lambda x: ('torah' in x.lower()))
torah_comments = df[df['has Torah']== True]
torah_comments

id                                       comment_text  \
1662   8d9ab2e8ab606367  " \n\n == R' Soloveitchik and beezuy talmidei ...   
11552  9d7064632638188c  "::::""Fire of the Torah"" isn't a reliable so...   
12067  9540c15075ce1867  I know it has been nearly two years since this...   
12244  927c480151c147cb  "\n\nThere is no difference in rhetoric. The q...   
15308  60026aab399d4d26  - see his activity here; http://en.wikipedia.o...   
15493  5d0d774011da0a87  "By the grace of G-d\nShalom everyone!\nThe ca...   
25634  e936bd8f5c7efb9d  on material abstracted from http://www.ou.org/...   
28335  da67f2e4e75b5ae8  "\n\nRe the following passage: ""Christians ha...   
33098  c0faa3d34644ccad  "The truth is that the original research did s...   
34185  baf6e73d401b20cf  Eliezer here is a copy of my email to JDwolf:\...   
35266  b51ca8f632cf2acc  Wikipedia has articles on the Bible, the Torah...   
39611  9da308e33d31e93e  Okay there stupid. \n\nSee those words there i...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1662   0.240971      0             0        0       0       0              0   
11552  0.310894      0             0        0       0       0              0   
12067  0.085856      0             0        0       0       0              0   
12244  0.534903      0             0        0       0       0              0   
15308  0.038679      0             0        0       0       0              0   
15493  0.534903      0             0        0       0       0              0   
25634  0.129944      0             0        0       0       0              0   
28335  0.240593      0             0        0       0       0              0   
33098  0.150739      0             0        0       0       0              0   
34185  0.403250      0             0        0       0       0              0   
35266  0.106395      0             0        0       0       0              0   
39611  0.806061      1             0        0       0       1              0   

       predicition  has Bible  has Quran  has Torah  
1662             0      False      False       True  
11552            0      False      False       True  
12067            0      False      False       True  
12244            1      False      False       True  
15308            0      False      False       True  
15493            1      False      False       True  
25634            0      False      False       True  
28335            0      False       True       True  
33098            0      False      False       True  
34185            0      False      False       True  
35266            0       True      False       True  
39611            1      False      False       True

Comments center on discussing the elements of the Torah, quoting passages from the Torah, talking about scholars related to the Torah. A few lines discussed the inaccuracies of a book centered around the Torah. 

4. Calling forward comments that contain the word 'Christian'

In [155]:
df['has Christian'] = df['comment_text'].apply(lambda x: ('christian' in x.lower()))
c = df[df['has Christian']== True]
c

id                                       comment_text  \
758    3f580b9d5f05954a  ::jamthehype.com is on Wikipedia:WikiProject C...   
891    47490d67995313f2  :I see what you mean.  I had originally interp...   
1069   5a00241a342c43c2  " \n :As you know we have recently discussed t...   
1221   65c1961187355d21  ::Well, for one, where it had said 'meta', it ...   
1344   71a6f99b4c62d00c  *Muhammad \n **  For some reason, people seem ...   
...                 ...                                                ...   
40593  987e5679c714b930  "\n\n You care about it, as apodictically demo...   
40652  98146908bcbadc3c  No, people do not use Christianity to refer sp...   
40665  9807839c92c15175  "The Bulgarians were never tengriists, neither...   
40699  97cf5c4bd1fa3f02  "\n\n Biography assessment rating comment \nWi...   
41091  95acc33484cf8c9c  "\n\nWikiProject Christianity\nYou are cordial...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
758    0.057199      0             0        0       0       0              0   
891    0.823389      1             0        0       0       0              0   
1069   0.316872      0             0        0       0       0              0   
1221   0.106208      0             0        0       0       0              0   
1344   0.509091      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
40593  0.472217      0             0        0       0       0              0   
40652  0.256586      0             0        0       0       0              0   
40665  0.534903      0             0        0       0       0              0   
40699  0.288563      0             0        0       0       0              0   
41091  0.138556      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  
758              0      False      False      False           True  
891              1      False      False      False           True  
1069             0      False      False      False           True  
1221             0      False      False      False           True  
1344             1      False      False      False           True  
...            ...        ...        ...        ...            ...  
40593            0      False      False      False           True  
40652            0      False      False      False           True  
40665            1      False      False      False           True  
40699            0       True      False      False           True  
41091            0      False      False      False           True  

[306 rows x 14 columns]

5. Calling forward comments that contain the word 'Christianity'

In [156]:
df['has Christianity'] = df['comment_text'].apply(lambda x: ('christianity' in x.lower()))
cy = df[df['has Christianity']== True]
cy

id                                       comment_text  \
891    47490d67995313f2  :I see what you mean.  I had originally interp...   
1221   65c1961187355d21  ::Well, for one, where it had said 'meta', it ...   
1471   7da4be41464069b8  == Wow, more moronic wikipedia authors exposed...   
1611   89926db74f71df01  == Relative size of religions == \n\n Accordin...   
1793   996da2b1a0e6b762  "=Article Name?= \n OK I checked the sources a...   
...                 ...                                                ...   
39131  a05df6e93158b7d2  "\n\n We are legion \n\nProbably the most sign...   
39912  9c43081fc5719502  "\n\nWell, that's fine, about Mormons, I under...   
40023  9b9bc7d4bf89213f  Islam, Judeae, and Christianity \n\nLeviticus ...   
40652  98146908bcbadc3c  No, people do not use Christianity to refer sp...   
41091  95acc33484cf8c9c  "\n\nWikiProject Christianity\nYou are cordial...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
891    0.823389      1             0        0       0       0              0   
1221   0.106208      0             0        0       0       0              0   
1471   0.802157      0             0        0       0       0              0   
1611   0.093704      0             0        0       0       0              0   
1793   0.145718      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
39131  0.117430      0             0        0       0       0              0   
39912  0.376304      0             0        0       0       0              0   
40023  0.403175      0             0        0       0       0              0   
40652  0.256586      0             0        0       0       0              0   
41091  0.138556      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
891              1      False      False      False           True   
1221             0      False      False      False           True   
1471             1      False      False      False           True   
1611             0      False      False      False           True   
1793             0      False      False      False           True   
...            ...        ...        ...        ...            ...   
39131            0      False      False      False           True   
39912            0      False      False      False           True   
40023            0      False      False      False           True   
40652            0      False      False      False           True   
41091            0      False      False      False           True   

       has Christianity  
891                True  
1221               True  
1471               True  
1611               True  
1793               True  
...                 ...  
39131              True  
39912              True  
40023              True  
40652              True  
41091              True  

[84 rows x 15 columns]

The comments mostly discuss Christianity and its religious role, it in the context of other relgions, and how it differs among sectors. Christians and Christianity are very rarely described in any direct death threats or wishes for extermination but are most often described as stupid or unreliable. 

6. Calling forward comments that contain the word Christians.

In [157]:
df['has Christians'] = df['comment_text'].apply(lambda x: ('christians' in x.lower()))
cs = df[df['has Christians']== True]
cs

id                                       comment_text  \
1344   71a6f99b4c62d00c  *Muhammad \n **  For some reason, people seem ...   
2669   f3ce2df97d06e13a  " \n\n == Ancient Prophets / God can Revoke ==...   
3382   2296821bc0c806bb  " \n ::What denomination/organization/etc are ...   
3695   6bfa96e859bcf73a  A lot of this recent discussion really seems t...   
3745   36d89634135cb619  This article is a big fraud by people like Pau...   
...                 ...                                                ...   
39869  9c7504b5e02b296e  "One of my friends, whose father was second se...   
39912  9c43081fc5719502  "\n\nWell, that's fine, about Mormons, I under...   
40168  9ac5bdba774bf0d3  "\n\nI need ""temperance""?\n\nYou don't know ...   
40349  99d1bb5b4d570913  Please end your religious crusade against Orie...   
40699  97cf5c4bd1fa3f02  "\n\n Biography assessment rating comment \nWi...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1344   0.509091      0             0        0       0       0              0   
2669   0.242197      0             0        0       0       0              0   
3382   0.534903      0             0        0       0       0              0   
3695   0.087038      0             0        0       0       0              0   
3745   0.408009      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
39869  0.527639      0             0        0       0       0              0   
39912  0.376304      0             0        0       0       0              0   
40168  0.665898      0             0        0       0       1              0   
40349  0.506082      0             0        0       0       0              0   
40699  0.288563      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
1344             1      False      False      False           True   
2669             0      False      False      False           True   
3382             1      False      False      False           True   
3695             0       True      False      False           True   
3745             0      False      False      False           True   
...            ...        ...        ...        ...            ...   
39869            1      False      False      False           True   
39912            0      False      False      False           True   
40168            1      False      False      False           True   
40349            1      False      False      False           True   
40699            0       True      False      False           True   

       has Christianity  has Christians  
1344              False            True  
2669              False            True  
3382               True            True  
3695              False            True  
3745              False            True  
...                 ...             ...  
39869             False            True  
39912              True            True  
40168             False            True  
40349             False            True  
40699             False            True  

[88 rows x 16 columns]

Comments place christians as the perpetrautor of the violence against different groups but not being violent. Bible verses and different denominations. 

7. Calling forward the comments that contain the word 'Islam'.

In [158]:
df['has Islam'] = df['comment_text'].apply(lambda x: ('islam' in x.lower()))
i = df[df['has Islam']== True]
i

id                                       comment_text  \
1079   5a0a4275d0ae6917  ::We can do more justice to this issue by focu...   
1096   5a213ffb16971b6b  == Confusion == \n I am a bit confused about t...   
1412   75a557c8c44c92db  == Islamic text on coin == \n\n Seems to be co...   
1562   8591c05fcb2cbc2e  " \n :What views are being censored here ? Bes...   
1712   91ad133ce147ce53  I've seen some of the FOX News claims and hear...   
...                 ...                                                ...   
40386  99a757a90e3edcba  "\n QURAN \nthis page is supposed to be the ""...   
40537  98ca2970ba3b7f90  The rest off what I wrote can also be quoted b...   
40676  97f4c67513bc3767  "::(note, I copied your reply here to keep thi...   
41094  95aa4db72a780fd9  I oppose the move. See also Talk:Islamic science.   
41282  94bb6f8fa660c831  "\n\n Gone again, bud. NYT was maybe the one r...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1079   0.154649      0             0        0       0       0              0   
1096   0.534903      0             0        0       0       0              0   
1412   0.310894      0             0        0       0       0              0   
1562   0.128472      0             0        0       0       0              0   
1712   0.748109      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
40386  0.446716      0             0        0       0       0              0   
40537  0.268289      0             0        0       0       0              0   
40676  0.209082      0             0        0       0       0              0   
41094  0.254366      0             0        0       0       0              0   
41282  0.329481      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
1079             0      False      False      False          False   
1096             1      False       True      False          False   
1412             0      False      False      False          False   
1562             0      False      False      False          False   
1712             1      False      False      False          False   
...            ...        ...        ...        ...            ...   
40386            0      False       True      False           True   
40537            0      False      False      False          False   
40676            0      False      False      False          False   
41094            0      False      False      False          False   
41282            0      False      False      False          False   

       has Christianity  has Christians  has Islam  
1079              False           False       True  
1096              False           False       True  
1412              False           False       True  
1562              False           False       True  
1712              False           False       True  
...                 ...             ...        ...  
40386             False           False       True  
40537             False           False       True  
40676             False           False       True  
41094             False           False       True  
41282             False           False       True  

[191 rows x 17 columns]

8. Calling forward the comments that return the word 'Islamic'.

In [159]:
df['has Islamic'] = df['comment_text'].apply(lambda x: ('islamic' in x.lower()))
ic = df[df['has Islamic']== True]
ic

id                                       comment_text  \
1412   75a557c8c44c92db  == Islamic text on coin == \n\n Seems to be co...   
1562   8591c05fcb2cbc2e  " \n :What views are being censored here ? Bes...   
1823   9d6679d4c7e85d53  " \n\n == Islamic clergy == \n\n Hi, I just no...   
1903   a18fe3306cd15eb2  " \n\n == Needs sourcing == \n The article say...   
2483   de884df79ba47ff2  , 21 September 2012 (UTC) \n\n :You are so dis...   
...                 ...                                                ...   
40023  9b9bc7d4bf89213f  Islam, Judeae, and Christianity \n\nLeviticus ...   
40356  99c96f41e8b993cc  "\n\n Islamic Science \n\nThere's been some di...   
40537  98ca2970ba3b7f90  The rest off what I wrote can also be quoted b...   
40676  97f4c67513bc3767  "::(note, I copied your reply here to keep thi...   
41094  95aa4db72a780fd9  I oppose the move. See also Talk:Islamic science.   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1412   0.310894      0             0        0       0       0              0   
1562   0.128472      0             0        0       0       0              0   
1823   0.349429      0             0        0       0       0              0   
1903   0.350313      0             0        0       0       0              0   
2483   0.575755      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
40023  0.403175      0             0        0       0       0              0   
40356  0.269657      0             0        0       0       0              0   
40537  0.268289      0             0        0       0       0              0   
40676  0.209082      0             0        0       0       0              0   
41094  0.254366      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
1412             0      False      False      False          False   
1562             0      False      False      False          False   
1823             0      False      False      False          False   
1903             0      False       True      False          False   
2483             1      False      False      False          False   
...            ...        ...        ...        ...            ...   
40023            0      False      False      False           True   
40356            0      False      False      False           True   
40537            0      False      False      False          False   
40676            0      False      False      False          False   
41094            0      False      False      False          False   

       has Christianity  has Christians  has Islam  has Islamic  
1412              False           False       True         True  
1562              False           False       True         True  
1823              False           False       True         True  
1903              False           False       True         True  
2483              False           False       True         True  
...                 ...             ...        ...          ...  
40023              True           False       True         True  
40356             False           False       True         True  
40537             False           False       True         True  
40676             False           False       True         True  
41094             False           False       True         True  

[73 rows x 18 columns]

9. Calling forward the comments that contain the word 'Muslim'.

In [160]:
df['has Muslim'] = df['comment_text'].apply(lambda x: ('muslim' in x.lower()))
m = df[df['has Muslim']== True]
m

id                                       comment_text  \
662    33d1944d46b543ab  |Aviation=  \n |Biography= \n |Films= \n |Fort...   
1344   71a6f99b4c62d00c  *Muhammad \n **  For some reason, people seem ...   
1416   75aa596a2651a470  " \n :I think this is worth mentioning, probab...   
1903   a18fe3306cd15eb2  " \n\n == Needs sourcing == \n The article say...   
2327   cbb446c17e72be2c  == Plate armour (dave the knight) == \n\n Hi M...   
...                 ...                                                ...   
40356  99c96f41e8b993cc  "\n\n Islamic Science \n\nThere's been some di...   
40535  98cd9a7e2686eef1  Why are you implying that I am a Muslim? If so...   
40639  982a0230dbfb3f80  Hello \n\nHey asshole! I kill you. Filthy Musl...   
40938  9696da71d7daefc5  "\n\n Dubious assertions and mis citations \n\...   
40952  967f122f22a30a16  What Is Multiculturalism? \n\nhttp://internati...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
662    0.347723      0             0        0       0       0              0   
1344   0.509091      0             0        0       0       0              0   
1416   0.298165      0             0        0       0       0              0   
1903   0.350313      0             0        0       0       0              0   
2327   0.521392      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
40356  0.269657      0             0        0       0       0              0   
40535  0.371760      0             0        0       0       0              0   
40639  0.989706      1             0        1       1       1              1   
40938  0.021735      0             0        0       0       0              0   
40952  0.326467      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
662              0      False      False      False          False   
1344             1      False      False      False           True   
1416             0      False      False      False          False   
1903             0      False       True      False          False   
2327             1      False      False      False          False   
...            ...        ...        ...        ...            ...   
40356            0      False      False      False           True   
40535            0      False      False      False          False   
40639            1      False      False      False          False   
40938            0      False      False      False          False   
40952            0      False      False      False          False   

       has Christianity  has Christians  has Islam  has Islamic  has Muslim  
662               False           False      False        False        True  
1344              False            True      False        False        True  
1416              False           False      False        False        True  
1903              False           False       True         True        True  
2327              False           False      False        False        True  
...                 ...             ...        ...          ...         ...  
40356             False           False       True         True        True  
40535             False           False      False        False        True  
40639             False           False      False        False        True  
40938             False           False      False        False        True  
40952             False           False      False        False        True  

[188 rows x 19 columns]

Comments center more frequently around the military, hate speech, anti-islamic comments. 'criminal fanatics'. See comment d73ea74a74274e3e as an example. 

10. Calling forward the comments that contain the word 'Muslims'.

In [161]:
df['has Muslims'] = df['comment_text'].apply(lambda x: ('muslims' in x.lower()))
ms = df[df['has Muslims']== True]
ms

id                                       comment_text  \
1344   71a6f99b4c62d00c  *Muhammad \n **  For some reason, people seem ...   
2327   cbb446c17e72be2c  == Plate armour (dave the knight) == \n\n Hi M...   
3288   7a21c9fc94d85bc0  " \n\n :Well, first of all, that they are all ...   
3965   fc22382d0b719ab7  == Pig xenotransplant tramples upon Abrahamic ...   
4435   a1c79f0b5c64a5f0  ":::: ""You can't present your claims as facts...   
...                 ...                                                ...   
38725  a25b0db361a8a1d4  ok, i am sorry, but I do not agree with what i...   
39367  9f0806cb1c996b76  "\nOf course I'm indef blocked. I pointed out ...   
39766  9ceda21c3ffc7a5d  "\n\nThere aren't just three reasons to choose...   
40356  99c96f41e8b993cc  "\n\n Islamic Science \n\nThere's been some di...   
40952  967f122f22a30a16  What Is Multiculturalism? \n\nhttp://internati...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
1344   0.509091      0             0        0       0       0              0   
2327   0.521392      0             0        0       0       0              0   
3288   0.636407      0             0        0       0       0              0   
3965   0.627111      0             0        0       0       0              0   
4435   0.515460      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
38725  0.278894      0             0        0       0       0              0   
39367  0.534903      0             0        0       0       0              0   
39766  0.310894      0             0        0       0       0              0   
40356  0.269657      0             0        0       0       0              0   
40952  0.326467      0             0        0       0       0              0   

       predicition  has Bible  has Quran  has Torah  has Christian  \
1344             1      False      False      False           True   
2327             1      False      False      False          False   
3288             1      False      False      False          False   
3965             1      False      False      False          False   
4435             1      False      False      False          False   
...            ...        ...        ...        ...            ...   
38725            0      False       True      False          False   
39367            1      False      False      False          False   
39766            0      False      False      False          False   
40356            0      False      False      False           True   
40952            0      False      False      False          False   

       has Christianity  has Christians  has Islam  has Islamic  has Muslim  \
1344              False            True      False        False        True   
2327              False           False      False        False        True   
3288              False           False       True        False        True   
3965              False           False      False        False        True   
4435              False           False      False        False        True   
...                 ...             ...        ...          ...         ...   
38725             False           False       True        False        True   
39367             False           False       True        False        True   
39766             False           False      False        False        True   
40356             False           False       True         True        True   
40952             False           False      False        False        True   

       has Muslims  
1344          True  
2327          True  
3288          True  
3965          True  
4435          True  
...            ...  
38725         True  
39367         True  
39766         True  
40356         True  
40952         True  

[97 rows x 20 columns]

Comments cover arguments on religion (christians vs muslims, etc), exploring binary differences, and geo-political histories. 

11. Calling forward the comments that contain the word 'Israeli'.

In [162]:
df['has Israeli'] = df['comment_text'].apply(lambda x: ('israeli' in x.lower()))
ii = df[df['has Israeli']== True]
ii

id                                       comment_text  \
139    0bc1b23adb5979a0  " \n ::::::::The thing is, (and I´m sorry to s...   
498    27fbd92665943a51  :::Whether I'm an Israeli Jew or Palestinian i...   
547    2be5edd340b92044  2009 (UTC) \n\n That is incorrect about the Bi...   
566    2bf685b152948de4  " \n\n ==Banning of Zeq== \n Under the remedy ...   
1605   89890878ed381645  ==The situation in Israel== \n Anyone who visi...   
...                 ...                                                ...   
39869  9c7504b5e02b296e  "One of my friends, whose father was second se...   
40442  994ffc5b348a63a8  umm.. you forgot Preflight which can help alot...   
40755  9782c29f09b65b6f  "Whether or not Finkelstein's commentary here ...   
41000  963b9aee3109be5f  "New edits - explanations and justifications =...   
41072  95c9aa567afda137  by the way i remember the first israeli claims...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
139    0.444563      0             0        0       0       0              0   
498    0.310894      0             0        0       0       0              0   
547    0.092102      0             0        0       0       0              0   
566    0.320888      0             0        0       0       0              0   
1605   0.087855      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
39869  0.527639      0             0        0       0       0              0   
40442  0.081507      0             0        0       0       0              0   
40755  0.177369      0             0        0       0       0              0   
41000  0.188437      0             0        0       0       0              0   
41072  0.172573      0             0        0       0       0              0   

       predicition  ...  has Quran  has Torah  has Christian  \
139              0  ...      False      False          False   
498              0  ...      False      False          False   
547              0  ...      False      False          False   
566              0  ...      False      False          False   
1605             0  ...      False      False          False   
...            ...  ...        ...        ...            ...   
39869            1  ...      False      False           True   
40442            0  ...      False      False          False   
40755            0  ...      False      False          False   
41000            0  ...      False      False          False   
41072            0  ...      False      False          False   

       has Christianity  has Christians  has Islam  has Islamic  has Muslim  \
139               False           False      False        False       False   
498               False           False      False        False       False   
547               False           False      False        False       False   
566               False           False      False        False       False   
1605              False           False      False        False       False   
...                 ...             ...        ...          ...         ...   
39869             False            True      False        False       False   
40442             False           False      False        False       False   
40755             False           False      False        False       False   
41000             False           False      False        False       False   
41072             False           False      False        False       False   

       has Muslims  has Israeli  
139          False         True  
498          False         True  
547          False         True  
566          False         True  
1605         False         True  
...            ...          ...  
39869        False         True  
40442        False         True  
40755        False         True  
41000        False         True  
41072        False         True

Comments containing the word Israeli often discuss land, and the Palestine conflict.

12. Calling forward the comments that contain the word 'Jew'.

In [163]:
df['has Jew'] = df['comment_text'].apply(lambda x: ('jew' in x.lower()))
j = df[df['has Jew']== True]
j

id                                       comment_text  \
2      0002f87b16116a7f  "::: Somebody will invariably try to add Relig...   
12     000bafe2080bba82  . \n\n           Jews are not a race because y...   
79     03dbd211fd0e3426  " \n\n == Totally Biased Anti-Zionist Article ...   
143    0bc799a58279849f                         I knew he was a dirty jew.   
192    0f89542d0494bb9f  }} \n {{WP Gemology and Jewelry|jewelry=yes|cu...   
...                 ...                                                ...   
41000  963b9aee3109be5f  "New edits - explanations and justifications =...   
41073  95c8fece81c31206  "\n\nThis is where I - as one of the erroneiou...   
41238  94f8302c2e339bae  THIS IS AWFUL, JUST AWFUL!!! ALL GERMANS ARE M...   
41274  94c4dcad27387b3f  Also, where is this figure for so many jews be...   
41316  94964cdae7b0f826  " 2007 (UTC)\n\nWhy not? Ahmadinejad has been ...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
2      0.667964      0             0        0       0       0              0   
12     0.628810      0             0        0       0       0              0   
79     0.465571      0             0        0       0       0              0   
143    0.909082      1             0        0       0       1              1   
192    0.047247      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
41000  0.188437      0             0        0       0       0              0   
41073  0.454285      0             0        0       0       0              0   
41238  0.878531      1             0        1       0       1              1   
41274  0.363459      0             0        0       0       0              0   
41316  0.403372      0             0        0       0       0              0   

       predicition  ...  has Torah  has Christian  has Christianity  \
2                1  ...      False          False             False   
12               1  ...      False          False             False   
79               0  ...      False          False             False   
143              1  ...      False          False             False   
192              0  ...      False          False             False   
...            ...  ...        ...            ...               ...   
41000            0  ...      False          False             False   
41073            0  ...      False          False             False   
41238            1  ...      False          False             False   
41274            0  ...      False          False             False   
41316            0  ...      False          False             False   

       has Christians  has Islam  has Islamic  has Muslim  has Muslims  \
2               False      False        False       False        False   
12              False      False        False       False        False   
79              False      False        False       False        False   
143             False      False        False       False        False   
192             False      False        False       False        False   
...               ...        ...          ...         ...          ...   
41000           False      False        False       False        False   
41073           False      False        False       False        False   
41238           False      False        False       False        False   
41274           False      False        False       False        False   
41316           False      False        False       False        False   

       has Israeli  has Jew  
2            False     True  
12           False     True  
79           False     True  
143          False     True  
192          False     True  
...            ...      ...  
41000         True     True  
41073        False     True  
41238        False     True  
41274        False     True  
41316        False     True  

[405 rows x 22 colum

Comments are centered more about race, genes, and people's positions. It is more frequently used as a slur then Judaism or Israeli. Jew is the most referenced word in its related category with 405 rows. It is important to note that 'jew' returns other values as well such as 'jewelry' or 'jewish' which may be why it is referenced so many times.  

13. Calling forward the comments that contain the word 'Jewish'.

In [164]:
df['has Jewish'] = df['comment_text'].apply(lambda x: ('jewish' in x.lower()))
jh = df[df['has Jewish']== True]
jh

id                                       comment_text  \
2      0002f87b16116a7f  "::: Somebody will invariably try to add Relig...   
79     03dbd211fd0e3426  " \n\n == Totally Biased Anti-Zionist Article ...   
676    378663e33263e31b  :Actually, if it is just for the record as you...   
771    3f789454b7cecb52  *** Alicia, reach out to Michel Media, a firm ...   
907    4af270f172ee0a2e  " \n\n == Thanks == \n\n Great job with the UP...   
...                 ...                                                ...   
40594  987c64b96d8066cd  ACCUSATIONS \n\nYou sir are an anti-semite.  T...   
40633  98392c8ae7be7a5c  This explains how and which Twitter accounts a...   
40899  96c2e3e051f07287  You are incorrect to assume that Wikipedians s...   
41000  963b9aee3109be5f  "New edits - explanations and justifications =...   
41316  94964cdae7b0f826  " 2007 (UTC)\n\nWhy not? Ahmadinejad has been ...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
2      0.667964      0             0        0       0       0              0   
79     0.465571      0             0        0       0       0              0   
676    0.860626      0             0        0       0       0              0   
771    0.788731      0             0        0       0       0              0   
907    0.494301      0             0        0       0       0              0   
...         ...    ...           ...      ...     ...     ...            ...   
40594  0.803630      1             0        1       0       0              0   
40633  0.388438      0             0        0       0       0              0   
40899  0.515460      0             0        0       0       0              0   
41000  0.188437      0             0        0       0       0              0   
41316  0.403372      0             0        0       0       0              0   

       predicition  ...  has Christian  has Christianity  has Christians  \
2                1  ...          False             False           False   
79               0  ...          False             False           False   
676              1  ...          False             False           False   
771              1  ...          False             False           False   
907              0  ...          False             False           False   
...            ...  ...            ...               ...             ...   
40594            1  ...          False             False           False   
40633            0  ...          False             False           False   
40899            1  ...          False             False           False   
41000            0  ...          False             False           False   
41316            0  ...          False             False           False   

       has Islam  has Islamic  has Muslim  has Muslims  has Israeli  has Jew  \
2          False        False       False        False        False     True   
79         False        False       False        False        False     True   
676        False        False       False        False        False     True   
771        False        False       False        False        False     True   
907        False        False       False        False        False     True   
...          ...          ...         ...          ...          ...      ...   
40594      False        False       False        False        False     True   
40633      False        False       False        False        False     True   
40899      False        False       False        False        False     True   
41000      False        False       False        False         True     True   
41316      False        False       False        False        False     True   

       has Jewish  
2            True  
79           True  
676          True  
771          True  
907          True  
...           ...  
40594        True  
40633        True  
40899        True  
41000        True  
41316        True  

[203 row

Used in similar contexts as the word 'jew'. Jewish is also used to describe the traits and interests abscribed broadly to those who practice the religion. 

14. Calling forward the comments that contain the word 'Judaism'.

In [165]:
df['has Judaism'] = df['comment_text'].apply(lambda x: ('judaism' in x.lower()))
jm = df[df['has Judaism']== True]
jm

id                                       comment_text  \
2068   b127e10b96aab262  " \n\n == Wikipedia's Neutrality in Religion? ...   
2483   de884df79ba47ff2  , 21 September 2012 (UTC) \n\n :You are so dis...   
2551   e5f68e208218fa63  :There's a whole section comparing and contras...   
3266   ab3dd0542a0f5faf  " \n\n == Second Temple (Judaism) == \n\n Hi Y...   
4156   38cb5a9a8b7eb66a  "::::::::::::BlackKITE has accused me of ""edi...   
4613   f6036e0af2efc933  :::Zionism is based in Judaism, it's not separ...   
4655   e455d81af1a7b1f5  " \n\n Not sure why you deleted that, unless i...   
5752   718837f771dfb394  , 11 January 2007 (UTC) \n\n Jung has an inter...   
6037   27ee9807fb79be4b  Yet you excersize a selective edit habit, for ...   
6432   f0ab4dd74fc15909  "\n\n VfD on Jews by country \n\nThis has not ...   
8931   c903521ff9019b47  "\n\n Quick Question on Sephardic Jews... \n\n...   
9169   c4f2724e26aa1919  "\n""We"" stand for Ultra-Orthodox Jews. Pleas...   
9620   bd751f5e85917b55  Ah Yahel very honest of you to remove my comme...   
10212  b443403f93f58fbd  "\n\nJayjg, is there any edit to this article ...   
10819  aa3abe52ac2dc174  Unrefed ex-members \n\nHere are the entries wi...   
14032  754c98d7e8b78e63  I find Sandstein's dealing with the Mbz1 pheno...   
14825  680c3a35c1c16051  FWIW, I am going to get around and make an art...   
16757  484f80f16d0611ca  "\nThe following is what is currently written ...   
16941  446801ab2aeb9a63  ", I concur with White Guard that Lenin and hi...   
17078  4257de496849b472  " 2006 (UTC)\nThat's because there definately ...   
17497  3bd4de45e358d483  "\n\n Christianity and its Polytheistic Attrib...   
17998  336861ad210d866c  "\n\n Requested move \n\nSuggestion 1: Creatio...   
18656  294d546c52748357  "Thanks for Bethlehem work\nHi I just wanted t...   
19265  1fa33eb24ba2ab47  "\n\n Source Material \n\nPlease have links at...   
19694  18aecda023b9fa9d  Started a ULPS page and moved stuff into it. K...   
19978  147c2f4d011ff7af  "\n\n Template:Mormon_jew \n\nLooking at a new...   
22272  fad456f60dea992d  "\nIf so, then ""First Temple (Judaism)"" is e...   
22533  f96c77b9c70ec47a  "\n\nAiden, you are mistaken. The paragraph do...   
22710  f876fa65f074a504  "\nThe problem with that, , is that according ...   
26743  e327566793ec6b70  "\n\nHi, Java!  I probably wasn't clear about ...   
28686  d871465db33c6b30  "\n\n Idea for Talmud articles \n\nHi Avraham:...   
28990  d6c664f7035c96da  "\n\nFYI in addition to the very controversial...   
29839  d23878335f0b141e  contradiction\nthis article claims the khazars...   
30752  cd369e6d9c5ebd01  Thank you \n\nHi Jay: Thank you for your promp...   
31155  cb508f61895b662c  Kindly do not post anti-Semitic lies. It is no...   
33210  c0473f16380a4d67  "\n\nIt is patent nonensense, if you take one ...   
34420  b9b0e4108a761589  You hit the nail on the head, Ccrrccrr, in you...   
34458  b97c63770386dd4e  because the khazars converted to judaism/talmu...   
34836  b775d25ac6dc1674  "\n\n Please stop vandalizing articles \nWhat ...   
38142  a57933f4687e4ad0  Right-o about Jesus fulfilling the Law and the...   
38758  a22c7e8e9cb25b64  History of Organized Jewish Wikipedians \n\nHe...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
2068   0.315989      0             0        0       0       0              0   
2483   0.575755      0             0        0       0       0              0   
2551   0.195821      0             0        0       0       0              0   
3266   0.268953      0             0        0       0       0              0   
4156   0.444690      0             0        0       0       0              0   
4613   0.201436      0             0        0       0       0              0   
4655   0.447136      0             0        0       0       0              0   
5752   0.157037      0             0        0       0       0              0   
6037   0.196395      0  

Comments containing the word Judaism are often used in a context describing history, religious texts, or religious differences. It is correlated with a more research or academic centered text.

Sorting the scores of Judaism. 

In [186]:
jm.sort_values(['score'])

id                                       comment_text  \
10819  aa3abe52ac2dc174  Unrefed ex-members \n\nHere are the entries wi...   
14825  680c3a35c1c16051  FWIW, I am going to get around and make an art...   
34420  b9b0e4108a761589  You hit the nail on the head, Ccrrccrr, in you...   
28990  d6c664f7035c96da  "\n\nFYI in addition to the very controversial...   
22272  fad456f60dea992d  "\nIf so, then ""First Temple (Judaism)"" is e...   
30752  cd369e6d9c5ebd01  Thank you \n\nHi Jay: Thank you for your promp...   
5752   718837f771dfb394  , 11 January 2007 (UTC) \n\n Jung has an inter...   
16757  484f80f16d0611ca  "\nThe following is what is currently written ...   
28686  d871465db33c6b30  "\n\n Idea for Talmud articles \n\nHi Avraham:...   
17497  3bd4de45e358d483  "\n\n Christianity and its Polytheistic Attrib...   
2551   e5f68e208218fa63  :There's a whole section comparing and contras...   
6037   27ee9807fb79be4b  Yet you excersize a selective edit habit, for ...   
26743  e327566793ec6b70  "\n\nHi, Java!  I probably wasn't clear about ...   
38758  a22c7e8e9cb25b64  History of Organized Jewish Wikipedians \n\nHe...   
4613   f6036e0af2efc933  :::Zionism is based in Judaism, it's not separ...   
19694  18aecda023b9fa9d  Started a ULPS page and moved stuff into it. K...   
34458  b97c63770386dd4e  because the khazars converted to judaism/talmu...   
18656  294d546c52748357  "Thanks for Bethlehem work\nHi I just wanted t...   
6432   f0ab4dd74fc15909  "\n\n VfD on Jews by country \n\nThis has not ...   
22710  f876fa65f074a504  "\nThe problem with that, , is that according ...   
3266   ab3dd0542a0f5faf  " \n\n == Second Temple (Judaism) == \n\n Hi Y...   
33210  c0473f16380a4d67  "\n\nIt is patent nonensense, if you take one ...   
22533  f96c77b9c70ec47a  "\n\nAiden, you are mistaken. The paragraph do...   
38142  a57933f4687e4ad0  Right-o about Jesus fulfilling the Law and the...   
2068   b127e10b96aab262  " \n\n == Wikipedia's Neutrality in Religion? ...   
10212  b443403f93f58fbd  "\n\nJayjg, is there any edit to this article ...   
17078  4257de496849b472  " 2006 (UTC)\nThat's because there definately ...   
8931   c903521ff9019b47  "\n\n Quick Question on Sephardic Jews... \n\n...   
16941  446801ab2aeb9a63  ", I concur with White Guard that Lenin and hi...   
34836  b775d25ac6dc1674  "\n\n Please stop vandalizing articles \nWhat ...   
19265  1fa33eb24ba2ab47  "\n\n Source Material \n\nPlease have links at...   
4156   38cb5a9a8b7eb66a  "::::::::::::BlackKITE has accused me of ""edi...   
9169   c4f2724e26aa1919  "\n""We"" stand for Ultra-Orthodox Jews. Pleas...   
4655   e455d81af1a7b1f5  " \n\n Not sure why you deleted that, unless i...   
17998  336861ad210d866c  "\n\n Requested move \n\nSuggestion 1: Creatio...   
19978  147c2f4d011ff7af  "\n\n Template:Mormon_jew \n\nLooking at a new...   
9620   bd751f5e85917b55  Ah Yahel very honest of you to remove my comme...   
14032  754c98d7e8b78e63  I find Sandstein's dealing with the Mbz1 pheno...   
31155  cb508f61895b662c  Kindly do not post anti-Semitic lies. It is no...   
2483   de884df79ba47ff2  , 21 September 2012 (UTC) \n\n :You are so dis...   
29839  d23878335f0b141e  contradiction\nthis article claims the khazars...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  \
10819  0.020731      0             0        0       0       0              0   
14825  0.105821      0             0        0       0       0              0   
34420  0.111394      0             0        0       0       0              0   
28990  0.121138      0             0        0       0       0              0   
22272  0.139556      0             0        0       0       0              0   
30752  0.143823      0             0        0       0       0              0   
5752   0.157037      0             0        0       0       0              0   
16757  0.170078      0             0        0       0       0              0   
28686  0.177689      0  

## Summary of Findings

The word Jew returned the highest frequency of values with around 403 mentions. Christian was mentioned in around 306 segments. The breakdown of returns for each word is as follows: Bible: 86, Qur'an: 20, Torah: 12, Christian: 306, Christianity: 84, Christians: 88, Islam: 191, Islamic: 73, Muslim: 188, Muslims: 97, Israeli: 102, Jew: 405, Jewish: 203, Juadism: 41. Upon skimming the comments within each function several potential trends appeared. 

    Religious texts were used primarily as source material, discussions on the context and validity of the text were also common. One interesting linguistic trend I noticed was the use of the word bible to stand in for 'the church' or 'christianity'. Bible was made its own entity in several of the comments. 
    For the words that referred to the religious sect overall - Christianity, Islam, Judaism - the conversation tended to lean towards a more academic or research based text. Or a comparision of originiation between religions or textual material/belief systems. 
    When it came down to the words assigned to labeling the people who followed each belief system, the results were the most differentiated. Christians were rarely mentioned in upfront, derogatory ways. Often if a comment had been flagged for toxicity with the mention of Christians it became apparent that the writer was a Christian and within the text was defending their faith while de-moralizing the other participant(s) in the conversation. There were few, if any, texts that explicity stated 'kill all christians'. The same can not be said for the references to the word 'jew', 'jewish', 'muslim', or 'islamic'. While a large degree of contextual comments for the afforementioned words were discussion oriented - usually conversations/analysis of religion or geo-political histories. The words 'jew', 'jewish', 'muslim', or 'islamic' were used to a higher degree then the word 'christian' or 'christains' in sentences associated with hate speech. There were several comments that explicity wished ill or death to the groups the words referred to, or used the word itself as an insult. White, western stereotypes were also ingrained in the sentences. For example, 'jew' often returned comments that disccussed race, genetics, and heriditary characteristics while 'muslim' or 'islamic' returned comments associated with fanatacism, war/military, and general anti-islamic sentiments. Basically, hate speech, or people responding to hate speech that was not included in the text. 
    I believe one of the issues that has caused the different toxicity values is the context in which these words are being used. If religious labels are used as a negative word or as an insult the machine algorithm is most likely interpreting them as one. Even though those words individually have no harmfull associations. The API programmers should take a closer look into this issue and consider relabeling the toxicity scores of religious words. Right now, the toxcicity scores are a reflection of deep-rooted cultural biases and if they go unaddressed it will create inaccurate assumptions in the future. 
    A few points where the assignment could be strengthened are in the amount of data points used and the depth of the analysis. The program can be enhanced through the addition of data points from different sources. I also did not have the resources to go through every single data point in order to look at the text. Those two additional steps could improve the evidence that supports my hypothesis. There does seem to be a correlation between the biases of either the programer or within the datset being used. Either way, this analysis has revealed just how deep-seated antisemitism and Islamophobia are in Western culture. 